dali/c_api/c_api.cc

```diff
@@ -105,20 +105,32 @@ void daliCreatePipeline(daliPipelineHandle* pipe_handle,
     pipe->SetQueueSizes(cpu_prefetch_queue_depth, gpu_prefetch_queue_depth);
   }
   pipe->Build();
-  pipe_handle->pipe = reinterpret_cast<void*>(pipe);
+  pipe_handle->pipe = pipe;
   pipe_handle->ws = new dali::DeviceWorkspace();
   CUDA_CALL(cudaStreamCreateWithFlags(&pipe_handle->copy_stream, cudaStreamNonBlocking));
 }
 
-void daliPrefetchUniform(daliPipelineHandle* pipe_handle, int queue_depth) {
-  dali::Pipeline* pipeline = reinterpret_cast<dali::Pipeline*>(pipe_handle->pipe);
+
+void daliDeserializeDefault(daliPipelineHandle *pipe_handle, const char *serialized_pipeline,
+                            int length) {
+  auto pipeline = std::make_unique<dali::Pipeline>(std::string(serialized_pipeline, length));
+  pipeline->Build();
+  CUDA_CALL(cudaStreamCreateWithFlags(&pipe_handle->copy_stream, cudaStreamNonBlocking));
+  pipe_handle->ws = new dali::DeviceWorkspace();
```

Comment:
If this throws (e.g. a bad_alloc, copy_stream will leak).
Suggested solution - create all resources first, in managed handles - and then only release them:
```C++
CUDAStream stream = CUDAStream::Create();
auto ws = make_unique<dali::DeviceWorkspace>()`
pipe_handle->ws = ws.release();
pipe_handle->copy_stream = stream.release();
pipe_handle->pipe = pipe.release();
```